File: src/main/java/com/sleepwalker/sleeplib/util/network/SimplePacketManager.java
package com.sleepwalker.sleeplib.util.network;

import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.network.NetworkDirection;
import net.minecraftforge.fml.network.NetworkEvent;
import net.minecraftforge.fml.network.NetworkRegistry;
import net.minecraftforge.fml.network.simple.SimpleChannel;

import javax.annotation.Nonnull;
import java.util.Optional;
import java.util.function.Supplier;

public class SimplePacketManager {

    @Nonnull protected final SimpleChannel channel;
    protected int nextMsgId;

    public SimplePacketManager(@Nonnull SimpleChannel channel) {
        this.channel = channel;
    }

    public SimplePacketManager(@Nonnull String version, @Nonnull ResourceLocation name) {
        channel = NetworkRegistry.newSimpleChannel(name, () -> version, version::equals, version::equals);
    }

    public <P extends IClientPacket> void rgCP(@Nonnull IPacketDecoder<P> decoder, @Nonnull Class<P> type) {
        rg(decoder, type, NetworkDirection.PLAY_TO_SERVER);
    }

    public <P extends IServerPacket> void rgSP(@Nonnull IPacketDecoder<P> decoder, @Nonnull Class<P> type){
        rg(decoder, type, NetworkDirection.PLAY_TO_CLIENT);
    }

    public <P extends INetworkPacket> void rg(@Nonnull IPacketDecoder<P> decoder, @Nonnull Class<P> type, @Nonnull NetworkDirection direction){
        channel.registerMessage(nextMsgId++, type, INetworkPacket::write, decoder::decode, this::handlePacket, Optional.of(direction));
    }

    protected <P extends INetworkPacket> void handlePacket(@Nonnull P packet, @Nonnull Supplier<NetworkEvent.Context> supplier){
        NetworkEvent.Context context = supplier.get();
        if(context.getDirection() == packet.getNetworkDirection()){
            context.enqueueWork(() -> packet.handle(context));
        }
        context.setPacketHandled(true);
    }

    @Nonnull
    public SimpleChannel getChannel() {
        return channel;
    }
}

File: src/main/kotlin/com/sleepwalker/sleeplib/util/network/SimplePacketManager.kt
package com.sleepwalker.sleeplib.util.network

import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.network.PacketBuffer
import net.minecraft.util.ResourceLocation
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn
import net.minecraftforge.fml.network.NetworkDirection
import net.minecraftforge.fml.network.NetworkEvent
import net.minecraftforge.fml.network.NetworkRegistry
import net.minecraftforge.fml.network.PacketDistributor
import net.minecraftforge.fml.network.simple.SimpleChannel
import java.util.*
import java.util.function.Supplier

open class SimplePacketManager(val channel: SimpleChannel) {

    constructor(version: String, name: ResourceLocation) : this(
        NetworkRegistry.newSimpleChannel(name, { version }, { version == it }, { version == it })
    )

    private var nextMsgId = 0

    fun <P : INetworkPacket> rg(type: Class<P>, decoder: (PacketBuffer) -> P, handler: (P, Supplier<NetworkEvent.Context>) -> Unit, direction: NetworkDirection) {
        channel.registerMessage(nextMsgId++, type,
            { packet, buffer -> packet.write(buffer) },
            { buffer -> decoder(buffer) },
            handler,
            Optional.of(direction)
        )
    }

    fun <P : INetworkPacket> rg(type: Class<P>, decoder: (PacketBuffer) -> P, direction: NetworkDirection) {
        rg(type, decoder, { packet, contextSupplier -> this.handlePacket(packet, contextSupplier) }, direction)
    }

    fun <P : IClientPacket> rgCP(decoder: (PacketBuffer) -> P, type: Class<P>) = rg(type, decoder,  NetworkDirection.PLAY_TO_SERVER)
    inline fun <reified P : IClientPacket> rgCP(noinline decoder: (PacketBuffer) -> P) = rgCP(decoder, P::class.java)
    fun <P : IServerPacket> rgSP(decoder: (PacketBuffer) -> P, type: Class<P>) = rg(type, decoder, NetworkDirection.PLAY_TO_CLIENT)
    inline fun <reified P : IServerPacket> rgSP(noinline decoder: (PacketBuffer) -> P) = rgSP(decoder, P::class.java)

    private fun <P : INetworkPacket> handlePacket(packet: P, contextSupplier: Supplier<NetworkEvent.Context>) {
        val context = contextSupplier.get()
        if (context.direction == packet.networkDirection) {
            context.enqueueWork { packet.handle(context) }
        }
        context.packetHandled = true
    }

    fun sendToPlayer(player: ServerPlayerEntity, packet: IServerPacket) {
        channel.send(PacketDistributor.PLAYER.with { player }, packet)
    }

    fun sendToAll(packet: IServerPacket) {
        channel.send(PacketDistributor.ALL.noArg(), packet)
    }

    @OnlyIn(Dist.CLIENT)
    fun sendToServer(packet: IClientPacket) {
        channel.sendToServer(packet)
    }
}